torch/_dynamo/variables/torch.py

```diff
@@ -888,14 +896,45 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
         if self.value.__name__ == "cond":
             # TODO(voz): Support fake tensor dispatch for recursive
             # ops - see torch/dispatch/_dispatcher.py
-            assert len(args) == 4
-            assert type(args[0]) in (
-                TensorVariable,
-                SymNodeVariable,
-                ConstantVariable,
-            ), str(
-                type(args[0])
-            )  # predicate
+            if len(args) != 4:
+                raise UserError(
+                    UserErrorType.DYNAMIC_CONTROL_FLOW,
```

Comment:
I mentioned this in someone else's PR, but I'm not sure why you don't have a error hierarchy, instead of splitting it into an enum + error name.

Reply:
I think we want only want to the very high level `UserError` to the users with a concise and actionable message. No need to expose detailed user error types, which is helpful for us but won't be useful for user. @avikchaudhuri mentioned similar point in a different comment.

Reply:
BTW, we just want a flat user-facing error, but we can still define internal error hierarchy to throw and catch specific errors and wrap those to the user-facing errors when needed.

Reply:
Err, concretely, I'm proposing DynamicControlFlowUserError(...) analog here. This doesn't seem overly detailed; it seems descriptive!

Reply:
From users perspective, having `DynamicControlFlowUserError` or `UserError.DynamicControlFlowError` doesn't make much different. I'm okay with ether way.

However, from developers perspective, I think we should keep either the user error hierarchy or user error type enum, but never do both. So I would propose keep using error type enum in this PR as it's already in use in multiple places in the code. If we agree to define a user error hierarchy going forward, we can have a follow-up PR migrating all use cases to it. 

Let me know your thoughts. CC: @ezyang @tugsbayasgalan @avikchaudhuri

Reply:
From the user's perspective, the most significant bit is whether something is a user error or an internal error. So that distinction is all the matters (along with the promise that all user errors have a high bar for actionable message, suggesting a fix / linking to an example).

I do think some kind of hierarchy is needed, otherwise we'll just end up with a "dumping ground" for all kinds of user error types without any organizing principle. Moreover, I like the idea of using subclasses because then we can add more fields specific to some kinds of user errors that are not applicable to others.

Now, what's the point of having an enum? To me, it's about having a flat list of error codes that we can use to collect analytics: how often do users hit some error code, how often do they search for a workaround given an error code, etc. The hierarchy is purely for managing our codebase.

So to have the best of both worlds, I'd suggest we move to a class-based hierarchy where the base class has a static error code (implemented as a monotonically increasing enum) along with any other common required fields. We print the error code as a prefix of the error message (something like `user error [code]`) to signal that it is a user error, but can otherwise take any other fields as well.

Reply:
Agree with most part. I think having user error hierarchy is sufficient to support both:
 1. better codebase management for error raising
 2. gather analytics. As long as the error can be aggregated, we will see things like 20% of users errors due to `DynamicControlFlowArgMismatch`, etc.

The error type enum (error code) can be optionally defined for each error in the hierarchy.

I may think differently about exposing error code to users, which is neither user understandable nor actionable w/o showing the meaning of it. The fact is, the meaning of it is already covered in the actionable error message.

Refactoring UserError to a user error hierarchy (like internal errors) will be addressed in a separate PR.

Reply:
Will start merging we agree on the actions.